src/object/srv_ec_aggregate.c

```diff
@@ -0,0 +1,1207 @@
+	daos_prop_t		*prop = NULL;
+	struct daos_prop_entry	*entry;
+	int			 rc = 0;
+
+	rc = ds_pool_iv_srv_hdl_fetch(agg_param->ap_pool_info.api_pool,
+				      &agg_param->ap_pool_info.api_poh_uuid,
+				      &agg_param->ap_pool_info.api_coh_uuid);
+	if (rc) {
+		D_ERROR("ds_pool_iv_srv_hdl_fetch failed: "DF_RC"\n",
+			DP_RC(rc));
+		goto out;
+	}
+
+	D_ALLOC_PTR(prop);
+	if (prop == NULL) {
+		D_ERROR("Property allocation failed\n");
+		rc = -DER_NOMEM;
+		goto out;
+	}
+
+	rc = ds_pool_iv_prop_fetch(agg_param->ap_pool_info.api_pool, prop);
+	if (rc) {
+		D_ERROR("ds_pool_iv_prop_fetch failed: "DF_RC"\n", DP_RC(rc));
+		goto out;
+	}
+
+	entry = daos_prop_entry_get(prop, DAOS_PROP_PO_SVC_LIST);
+	D_ASSERT(entry != NULL);
+	agg_param->ap_pool_info.api_svc_list =
+					(d_rank_list_t *)entry->dpe_val_ptr;
+
+	ABT_eventual_set(agg_param->ap_pool_info.api_eventual,
+			 (void *)&rc, sizeof(rc));
+out:
+	D_FREE(prop);
+}
+
+/* Iterates entire VOS. Invokes nested iterator to recurse through trees
+ * for all objects meeting the criteria: object is EC, and this target is
+ * leader.
+ */
+static int
+agg_iterate_all(struct ds_cont_child *cont)
+{
+	vos_iter_param_t	 iter_param = { 0 };
+	struct vos_iter_anchors  anchors = { 0 };
+	struct ec_agg_param	 agg_param = { 0 };
+	daos_handle_t		 ph = DAOS_HDL_INVAL;
+	int			*status;
+	int			 rc = 0;
+
+	uuid_copy(agg_param.ap_pool_info.api_pool_uuid,
+		  cont->sc_pool->spc_uuid);
+	uuid_copy(agg_param.ap_pool_info.api_cont_uuid, cont->sc_uuid);
+
+	agg_param.ap_pool_info.api_pool_version =
+		cont->sc_pool->spc_pool->sp_map_version;
+	agg_param.ap_pool_info.api_pool = cont->sc_pool->spc_pool;
+	agg_param.ap_cont_handle	= cont->sc_hdl;
+
+	rc = ABT_eventual_create(sizeof(*status),
+				 &agg_param.ap_pool_info.api_eventual);
+	if (rc != ABT_SUCCESS)
+		return dss_abterr2der(rc);
+	rc = dss_ult_create(agg_iv_ult, &agg_param, DSS_ULT_POOL_SRV, 0, 0,
+			    NULL);
+	if (rc)
+		goto out;
+	rc = ABT_eventual_wait(agg_param.ap_pool_info.api_eventual,
+			       (void **)&status);
+	if (rc != ABT_SUCCESS) {
+		rc = dss_abterr2der(rc);
+		goto out;
+	}
+	if (*status != 0) {
+		rc = *status;
+		goto out;
+	}
+
+	rc = dsc_pool_open(agg_param.ap_pool_info.api_pool_uuid,
+			   agg_param.ap_pool_info.api_poh_uuid, 0, NULL,
+			   agg_param.ap_pool_info.api_pool->sp_map,
+			   agg_param.ap_pool_info.api_svc_list, &ph);
+	if (rc) {
+		D_ERROR("dsc_pool_open failed: "DF_RC"\n", DP_RC(rc));
+		goto out;
+	}
+
+	rc = dsc_cont_open(ph, agg_param.ap_pool_info.api_cont_uuid,
+			   agg_param.ap_pool_info.api_coh_uuid, 0,
+			   &agg_param.ap_pool_info.api_cont_hdl);
+	if (rc) {
+		D_ERROR("dsc_cont_open failed: "DF_RC"\n", DP_RC(rc));
+		goto out;
+	}
+
+	iter_param.ip_hdl		= cont->sc_hdl;
+	iter_param.ip_epr.epr_lo	= 0ULL;
+	iter_param.ip_epr.epr_hi	= DAOS_EPOCH_MAX;
+
+	rc = vos_iterate(&iter_param, VOS_ITER_OBJ, false, &anchors,
+			 agg_iter_obj_pre_cb, NULL, &agg_param, NULL);
```

Comment:
Iteration can't hog CPU for too long, it has to yield from time to time, also, re-probe could be required when iteration back from yield. (The re-probe is handled by vos_iterate(), but iterate callback needs to inform vos_iterate() if it yield).